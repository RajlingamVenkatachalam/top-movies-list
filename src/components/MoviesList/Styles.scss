@import "../variables.scss";

.main-panel {
    display:flex;
    height: 100%;
    width: 100%;
    flex-direction: column;
}

@media screen and (max-width: 1500px) {
    .main-panel-with-details {
        display:flex;
        height: 100%;
        width: 65%;
        flex-direction: column;
    }
}

@media screen and (max-width: 600px) {
    .main-panel-with-details {
        display:flex;
        height: 100%;
        width: 100%;
        flex-direction: column;
    }
}

.movie-list {
    margin: 10px;
}

.order-select-holder {
    text-align: left;
    margin: 20px;
    position: relative;
}

.movie-row {
    background-color: $bgColor;
    box-shadow: 10px 10px 5px $shadowColor;
    font-family: $fontFamily;
    cursor: pointer;

    .image {
        width: 100px;
    }
    .title{
        text-align: left;
        font-weight: bolder;
    }
}

.movie-row__selected {
    background-color: $shadowColor;
    font-weight: bolder;
    cursor: pointer;
    font-family: $fontFamily;
    text-align: left;
}
