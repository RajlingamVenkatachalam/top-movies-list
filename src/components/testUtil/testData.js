export const orderSelect = [{
        "label": "Release Date",
        "valueToOrderBy": "releaseDate"
      }, {
        "label": "Rank",
        "valueToOrderBy": "rank"
      }];

export const moviesList = [
    {
        "id": 2,
        "type": "poster",
        "rank": 2,
        "synopsis": "Epic tale in which an intrepid archaeologist tries to beat a band of Nazis to a unique religious relic which is central to their plans for world domination. Battling against a snake phobia and a vengeful ex-girlfriend, Indiana Jones is in constant peril, making hair's-breadth escapes at every turn in this celebration of the innocent adventure movies of an earlier era.",
        "title": "Raiders of the Lost Ark",
        "imageUrl": "https://preview.ibb.co/fn5Xyp/raiders.jpg",
        "releaseDate": 1981
      }, {
        "id": 4,
        "type": "poster",
        "rank": 1,
        "synopsis": "Deckard (Harrison Ford) is forced by the police Boss (M. Emmet Walsh) to continue his old job as Replicant Hunter. His assignment: eliminate four escaped Replicants from the colonies who have returned to Earth. Before starting the job, Deckard goes to the Tyrell Corporation and he meets Rachel (Sean Young), a Replicant girl he falls in love with.",
        "title": "Blade Runner",
        "imageUrl": "https://preview.ibb.co/h9434U/blade_Runner.jpg",
        "releaseDate": 1982
      }, {
        "id": 3,
        "type": "poster",
        "rank": 4,
        "synopsis": "Disguised as a human, a cyborg assassin known as a Terminator (Arnold Schwarzenegger) travels from 2029 to 1984 to kill Sarah Connor (Linda Hamilton). Sent to protect Sarah is Kyle Reese (Michael Biehn), who divulges the coming of Skynet, an artificial intelligence system that will spark a nuclear holocaust. Sarah is targeted because Skynet knows that her unborn son will lead the fight against them. With the virtually unstoppable Terminator in hot pursuit, she and Kyle attempt to escape.",
        "title": "The Terminator",
        "imageUrl": "https://preview.ibb.co/caykJp/terminator.jpg",
        "releaseDate": 1984
      }
];
