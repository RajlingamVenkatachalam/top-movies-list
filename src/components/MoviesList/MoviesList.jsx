import React from "react";
import PropTypes from "prop-type";
import InputLabel from "@material-ui/core/InputLabel";
import Select from "@material-ui/core/Select";
import MenuItem from "@material-ui/core/MenuItem";
import { sortMovies } from "../util";
import "./Styles.scss";

const MoviesList = ({ selectedMovieId, handleMovieSelection, moviesList, orderSelect }) => {
    const [selectedOrder, setSelectedOrder] = React.useState("");

    const handleOrderSelectChange = (event) => {
        setSelectedOrder(event.target.value);
    }

    return(
        <div className={`main-panel${selectedMovieId!==undefined ? "-with-details":""}`} >
            <div className="order-select-holder">
                <InputLabel  id="order-select-title">Sort By</InputLabel>
                <Select
                style={{width:"200px"}}
                  id="order-select"
                  value={selectedOrder}
                  onChange={handleOrderSelectChange}
                >
                    {
                        orderSelect.map((item, index)=><MenuItem key={index} value={item.valueToOrderBy}>{item.label}</MenuItem>)
                    }
                </Select>
            </div>
                <table className="movie-list" style={{position:"relative", height:"600px", overflow:"scroll"}}>
                    {
                       sortMovies(moviesList, selectedOrder).map(movie => 
                           <tr id={movie.id} className={`movie-row${movie.id===selectedMovieId ? "__selected":""}`} key={movie.id} onClick={()=> handleMovieSelection(movie)}>
                               <td className="image"><img src={movie.imageUrl} width="100" height="100" /></td>
                               <td className="title">{movie.title}</td>
                           </tr>
                       )                        
                    }
                </table>
        </div>
    )
}


MoviesList.propTypes = {
    selectedMovieId: PropTypes.number,
    handleMovieSelection: PropTypes.func
}

export default MoviesList;