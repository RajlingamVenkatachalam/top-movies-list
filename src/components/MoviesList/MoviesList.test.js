import React from "react";
import renderer from "react-test-renderer";
import MoviesList from "./MoviesList";
import { moviesList, orderSelect } from "../testUtil/testData";
import Select from "@material-ui/core/Select";

describe("test MoviesList", () => {
    it("should display the Detail without any crash", () =>{
        const onChangeMovieSelection = jest.fn();

        let component = renderer.create(<MoviesList  selectedMovieId={""} handleMovieSelection={onChangeMovieSelection}
            moviesList={moviesList} orderSelect={orderSelect} />);
        expect(component.toJSON()).toMatchSnapshot();
    });

    it("should call onChangeMovieSelection on Movie selection Change ", () =>{
        const onChangeMovieSelection = jest.fn();
        
        let component = renderer.create(<MoviesList  selectedMovieId={""} handleMovieSelection={onChangeMovieSelection}
            moviesList={moviesList} orderSelect={orderSelect} />);
        const instance = component.root;
        instance.findAllByType("tr")[0].props.onClick();
        expect(onChangeMovieSelection).toHaveBeenCalledWith(moviesList[0]);

        instance.findAllByType("tr")[1].props.onClick();
        expect(onChangeMovieSelection).toHaveBeenCalledWith(moviesList[1]);
    });

    it("should order the movies based on the selected sorting option to rank ", () =>{
        const onChangeMovieSelection = jest.fn();
        
        let component = renderer.create(<MoviesList  selectedMovieId={""} handleMovieSelection={onChangeMovieSelection}
            moviesList={moviesList} orderSelect={orderSelect} />);
        const instance = component.root;
        instance.findByType(Select).props.onChange({target: { value : "rank"}});

        const allMovies = instance.findAllByType("tr");
        expect(allMovies[0].props.id).toBe(4);
        expect(allMovies[1].props.id).toBe(2);
        expect(allMovies[2].props.id).toBe(3);
    });

    it("should order the movies based on the selected sorting option to releaseDate", () =>{
        const onChangeMovieSelection = jest.fn();
        
        let component = renderer.create(<MoviesList  selectedMovieId={""} handleMovieSelection={onChangeMovieSelection}
            moviesList={moviesList} orderSelect={orderSelect} />);
        const instance = component.root;
        instance.findByType(Select).props.onChange({target: { value : "releaseDate"}});

        const allMovies = instance.findAllByType("tr");
        expect(allMovies[0].props.id).toBe(2);
        expect(allMovies[1].props.id).toBe(4);
        expect(allMovies[2].props.id).toBe(3);
    });

});