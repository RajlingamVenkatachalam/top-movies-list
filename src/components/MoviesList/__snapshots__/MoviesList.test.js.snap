// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test MoviesList should display the Detail without any crash 1`] = `
<div
  className="main-panel-with-details"
>
  <div
    className="order-select-holder"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated"
      id="order-select-title"
    >
      Sort By
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline"
      onClick={[Function]}
      style={
        Object {
          "width": "200px",
        }
      }
    >
      <div
        aria-haspopup="listbox"
        aria-labelledby="order-select"
        className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
        id="order-select"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="button"
        tabIndex={0}
      >
        <span
          dangerouslySetInnerHTML={
            Object {
              "__html": "&#8203;",
            }
          }
        />
      </div>
      <input
        aria-hidden={true}
        className="MuiSelect-nativeInput"
        onAnimationStart={[Function]}
        onChange={[Function]}
        tabIndex={-1}
        value=""
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSelect-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
  <table
    className="movie-list"
    style={
      Object {
        "height": "600px",
        "overflow": "scroll",
        "position": "relative",
      }
    }
  >
    <tr
      className="movie-row"
      id={2}
      onClick={[Function]}
    >
      <td
        className="image"
      >
        <img
          height="100"
          src="https://preview.ibb.co/fn5Xyp/raiders.jpg"
          width="100"
        />
      </td>
      <td
        className="title"
      >
        Raiders of the Lost Ark
      </td>
    </tr>
    <tr
      className="movie-row"
      id={4}
      onClick={[Function]}
    >
      <td
        className="image"
      >
        <img
          height="100"
          src="https://preview.ibb.co/h9434U/blade_Runner.jpg"
          width="100"
        />
      </td>
      <td
        className="title"
      >
        Blade Runner
      </td>
    </tr>
    <tr
      className="movie-row"
      id={3}
      onClick={[Function]}
    >
      <td
        className="image"
      >
        <img
          height="100"
          src="https://preview.ibb.co/caykJp/terminator.jpg"
          width="100"
        />
      </td>
      <td
        className="title"
      >
        The Terminator
      </td>
    </tr>
  </table>
</div>
`;
